cage.c

```diff
@@ -146,13 +146,18 @@ static bool
 drop_permissions(void)
 {
 	if (getuid() != geteuid() || getgid() != getegid()) {
-		if (setuid(getuid()) != 0 || setgid(getgid()) != 0) {
-			wlr_log(WLR_ERROR, "Unable to drop root, refusing to start");
+		// Set the gid and uid in the correct order.
+		if (setgid(getgid()) != 0) {
+			wlr_log(WLR_ERROR, "Unable to drop root group, refusing to start");
+			return false;
+		}
+		if (setuid(getuid()) != 0) {
```

Comment:
Is there a reason for splitting this out over two branches? If not, I prefer the condensed one with the or operator (`||`) and the single log `wlr_log(WLR_ERROR, "Unable to drop root, refusing to start");`.

Reply:
Do keep the comment, though. That's a good one to make the ordering explicit :)

Reply:
The reference guide has it split up https://wiki.sei.cmu.edu/confluence/display/c/POS36-C.+Observe+correct+revocation+order+while+relinquishing+privileges

I don't really know much of POSIX C, so are you sure that the shorter version is equivalent to this? Does the conditinal branch have explicit guarantee's over what executes first? The current way is explicit in ordering.

Also see: https://github.com/swaywm/sway/pull/911

Reply:
I changed the error message to `Unable to drop root group, refusing to start`. Then I build a version of cage with the current PR and one with a single branch with the conditional. 

```
diff build1/cage build2/cage
Binary files buil1/cage and build2/cage differ
```

Reply:
So, I think with short-circuit evaluation in the OR conditional if the first statement resolves to true -- ie setgid() fails -- the program ignores the second statement -- which means the uid is not dropped. This isn't a problem here -- since the function fails anyway --, but this sort of oversight could be if no one was paying attention -- especially since this is a security-critical section of code.

The longer form is more explicit and more readable than the concise one IMO.

Please correct me if I'm wrong.

Reply:
Thanks for the extensive investigation @travankor, it's appreciated!

> are you sure that the shorter version is equivalent to this?

Yes, the ordering of the `||` operator is such that the left side executes first and unless it is true, the second operand is executed after. Here we're testing whether `setgid` fails and if that is true, there is no need to execute `setuid`. If it's false (hence, `setgid` succeeds) then we'll execute `setuid`.

> I changed the error message to Unable to drop root group, refusing to start. Then I build a version of cage with the current PR and one with a single branch with the conditional.

I don't think this is a valid comparison. Different structures are compiled differently due to optimizations, so you can't 1:1 compare two compiled binaries without turning off optimizations. Even then, it's probably best to inspect the generated assembly rather than the binaries.

> The longer form is more explicit and more readable than the concise one IMO.

Sure :) It is definitely more explicit; whether it is more readable is in the eye of the programmer. Thanks for the investigation and the PR!